### PR TITLE
fix: Allow tickets to have same name as deleted ones

### LINKED ISSUE #6621: Cannot create ticket with name "Business Ticket" if former ticket with name "Business Ticket" was deleted
Cannot create ticket with name "Business Ticket" if former ticket with name "Business Ticket" was deleted. Error message appears.

### PR DESCRIPTION
**Pull request by codedsun**
Fixes #6621 : Modified unique constraints in ticket model

#### Short description of what this resolves:
When deleting event the name was not changed of ticket which resulted in unique key constraint while creating another ticket with same name.

#### Changes proposed in this pull request:

- unique constraint changed to name, event and deleted_at

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [x] The unit tests pass locally with my changes 
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [x] All the functions created/modified in this PR contain relevant docstrings.

### CODE REVIEW

**Comment by codedsun** [comment]
@iamareebjamal I have tested this on my local with every case
1. When a non-corganizer deletes
2. deleting through the frontend and also the API.

Working fine for me 👍

File: app/api/tickets.py [hunk 1]
```diff
@@ -182,6 +182,11 @@ def before_update_object(self, ticket, data, view_kwargs):
                 raise UnprocessableEntity(
                     {'event_id': ticket.event.id}, "Event having paid ticket must have a payment method")
 
+        if ticket.deleted_at != data.get('deleted_at'):
+            if not has_access('is_coorganizer', event_id=ticket.event_id):
+                raise ForbiddenException({'source': ''}, 'Co-organizer access is required.')
```
Comment (iamareebjamal) [inline app/api/tickets.py]: This is already checked below in decorators

File: app/api/tickets.py [hunk 2]
```diff
@@ -182,6 +182,11 @@ def before_update_object(self, ticket, data, view_kwargs):
                 raise UnprocessableEntity(
                     {'event_id': ticket.event.id}, "Event having paid ticket must have a payment method")
 
+        if ticket.deleted_at != data.get('deleted_at'):
```
Comment (iamareebjamal) [inline app/api/tickets.py]: This will be true when restoring tickets from deleted as well

File: app/api/tickets.py (hunk 1 above)
Comment (codedsun) [reply app/api/tickets.py]: Ok will remove

File: app/api/tickets.py (hunk 2 above)
Comment (codedsun) [reply app/api/tickets.py]: So should I remove this? and just add the line for append?

Comment (codedsun) [reply app/api/tickets.py]: We should add a check when data.get('deleted_at') is null correct?

Comment (iamareebjamal) [reply app/api/tickets.py]: Check if there's a feature to restore deleted tickets in admin panel. If there's not, then it won't need to be changed

Comment (codedsun) [reply app/api/tickets.py]: No, there's nothing as such @iamareebjamal

**Comment by iamareebjamal** [comment]
OK, then create a script to do the same to existing deleted tickets

**Comment by codedsun** [comment]
@iamareebjam scheduled job ? Please describe a little more. Thanks

**Comment by iamareebjamal** [comment]
I'm thinking if we change the unique constraint from name to (name, deleted_at), this will be solved in much better way. Please try doing that instead

**Comment by codedsun** [comment]
@iamareebjamal Why dont we have a check of state in the system? If we have a state like 
1 - Active 
2- Inactive
3 - Deleted

Can't this solve the problem?

**Comment by iamareebjamal** [comment]
If a patient has sore throat, giving it medicine can fix the problem as well. Or cutting the patient's throat will also end the sore throat, we recommend medicines. Adding complexities like active inactive deleted pending just_about_to_start almost_ready pinky_promise is not a solution

**Comment by codedsun** [comment]
I will do this today!

**Comment by codedsun** [comment]
@iamareebjamal Currently the unique constraint are applied to name, eventId. Shal this be changed to name, deletedate and eventId.?

**Comment by codedsun** [comment]
@iamareebjamal While adding the constraint in the __table_args__ and applying migrations, I am not able to add new tickets. I am getting error
```INFO:werkzeug:127.0.0.1 - - [03/Dec/2019 20:49:39] "POST /v1/tickets HTTP/1.1" 405 -```

Screenshot 2019-12-03 at 8 51 01 PM

**Comment by prateekj117** [comment]
@codesun Update your branch. Check Gitter.

**Comment by codedsun** [comment]
@iamareebjamal @prateekj117 Please check this!

**Comment by prateekj117** [comment]
@codedsun Also make a migration to handle the already existing deleted tickets.

**Comment by codedsun** [comment]
@prateekj117 How do i do this? Please explain more

**Comment by prateekj117** [comment]
@codedsun The approach you were following before was correct. Unique Constraint was already there.

**Comment by codedsun** [comment]
@prateekj117 Unique constraint was only for name and event-id. Not for deleted-at, hence the issue was caused.

**Comment by prateekj117** [comment]
@codedsun Ohh, ya. I think this approach is also right. In this case, you won't even need to handle the already existing cases.

**Comment by codedsun** [comment]
> @codedsun Also make a migration to handle the case of already existing deleted tickets.

@prateekj117 Guide me for this

**Comment by iamareebjamal** [comment]
@codedsun Have you tested and confirmed that it is working?

**Comment by prateekj117** [comment]
@codedsun Not required now. Though, I was asking for something similar to this: https://github.com/fossasia/open-event-server/pull/6386

File: app/api/tickets.py [hunk 3]
```diff
@@ -17,7 +17,7 @@
 from app.models.ticket import Ticket, TicketTag, ticket_tags_table
 from app.models.event import Event
 from app.models.ticket_holder import TicketHolder
-from app.api.helpers.exceptions import ConflictException, MethodNotAllowed, UnprocessableEntity
```
Comment (iamareebjamal) [inline app/api/tickets.py]: Revert

**Comment by iamareebjamal** [comment]
You've just added constraint in model. Generate the migration as well

**Comment by codedsun** [comment]
@iamareebjamal 
Created a ticket named Paid ticket

image

Deleted and created the ticket with same name, working fine
image

File: app/api/tickets.py (hunk 3 above)
Comment (codedsun) [reply app/api/tickets.py]: Added!

**Comment by codedsun** [comment]
@iamareebjamal Completed, check now!

**Comment by iamareebjamal** [comment]
Also check that you cannot create "Paid Ticket" if it is not deleted

**Comment by codedsun** [comment]
@iamareebjamal Throws error of Ticket Already exist, working fine :)

image

image

**Pull request merged**